class-instant-articles-amp-markup.php

```diff
@@ -0,0 +1,163 @@
+
+		return self::$settings;
+	 }
+	/**
+	 * Checks if the AMP markup is enabled
+	 *
+	 * @return bool true if markup is enabled
+	 * @since 4.0
+	 */
+	static function is_markup_enabled() {
+		$settings = self::get_settings();
+
+		return
+			isset( $settings[self::SETTING_AMP_MARKUP] )
+			? (bool) $settings[self::SETTING_AMP_MARKUP]
+			: false;
+	}
+	/**
+	 * Adds the meta tag for AMP Markup if enabled
+	 *
+	 * @since 4.0
+	 */
+	static function inject_link_rel() {
+
+		if (!self::is_markup_enabled())
+			return;
+
+		// Transform the post to an Instant Article.
+		$adapter = new Instant_Articles_Post( get_post() );
+
+		if (!$adapter->should_submit_post())
+			return;
+
+		$url = $adapter->get_canonical_url();
+		$url = add_query_arg(self::QUERY_ARG, '1', $url );
+
+		echo '<link rel="amphtml" href="' . $url . '">';
+
+	}
+
+	/**
+	 * Generates the AMP markup if post has amp_markup
+	 *
+	 * NOTE: side-effect: function calls die() in the end.
+	 * @since 4.0
+	 */
+	static function markup_version( ) {
+		if (!(isset($_GET[ self::QUERY_ARG ]) && $_GET[ self::QUERY_ARG ]))
+			return;
+
+		$settings = self::get_settings();
+
+		if (!self::is_markup_enabled())
+			return;
+
+		$has_stylesheet =
+			isset($settings[self::SETTING_STYLE])
+			? self::validate_json($settings[self::SETTING_STYLE])
+			: false;
+
+		$properties = array();
+
+		$download_media =
+			isset($settings[self::SETTING_DL_MEDIA])
+			? (bool) $settings[self::SETTING_DL_MEDIA]
+			: false;
+
+		$properties[AMPArticle::ENABLE_DOWNLOAD_FOR_MEDIA_SIZING_KEY] = $download_media;
+
+		if ($has_stylesheet) {
+			$properties[AMPArticle::OVERRIDE_STYLES_KEY] =
+				json_decode($settings[self::SETTING_STYLE], true);
+		}
+
+		$post = get_post();
+
+		// This array will hold the image sizes
+		$media_sizes = array();
+
+		// Get all children with mime type image that are attached to the posts
+		// NOTE: this will not get images that are not hosted in the WP
+		$image_children = get_children('post_type=attachment&post_mime_type=image&post_parent=' . $post->ID );
+
+		foreach ($image_children as $img_id => $img) {
+			$meta = wp_get_attachment_metadata($img_id);
+
+			// Removes the file name from the URL
+			$url_chunks = explode('/', $img->guid);
+			array_pop($url_chunks);
+			$base_image_url = implode('/', $url_chunks) . '/';
+
+			// This is the uploaded original file
+			$media_sizes[$img->guid] = array($meta['width'], $meta['height']);
+
+			// These are the possible redimensions
+			foreach ($meta['sizes'] as $size) {
+				$size_url = $base_image_url.$size['file'];
+				$media_sizes[$size_url] = array($size['width'], $size['height']);
+			}
+		}
+
+		$properties[AMPArticle::MEDIA_SIZES_KEY] = $media_sizes;
+
+		// Transform the post to an Instant Article.
+		$adapter = new Instant_Articles_Post( $post );
+		$article = $adapter->to_instant_article();
+		$article_html = $article->render();
+		$amp = AMPArticle::create($article_html, $properties);
```

Comment:
Lets separate this AMP generation into a filter:
"get_amp_content"

Then the die call should be from an action